Proposing macro based shorthand.

andrewgu opened the pull request:
Note: This does not compile. This is another API change proposal.

Factors in suggestion to split out the setting into individual interface files. Also shows a macro based shorthand system that is more concise (also less explicit) than default handler approach. See MCM_ExampleTestHarness.uc for how it looks in practice.

andrewgu committed:
Proposing macro based shorthand.

andrewgu commented:
@Superd22 ^

andrewgu committed:
Added injection of save/cancel/reset handlers.

Superd22 commented:
That's perfect. :)
2 questions though, 
1. How do you see the SaveSetting implementation ? 
2. How do you wish to distribute the workload for the actual implementation (of every changes) ?

BlueRaja commented on ModConfigMenu/ModConfigMenu/Src/ModConfigMenuAPI/Classes/MCM_API_Instance.uc:
```diff
@@ -1,5 +1,13 @@
+enum eSettingEventMode
+{
+    // Callbacks are triggered when the user manipulates the UI element. Use SetSaveHandler and SetCancelHandler to save/abort.
+    eSettingEventMode_Immediate,
```
I _really_ don't like that this enum.  It's going to make the behavior of the API more inconsistent and bug-prone, for both us and users of the API.

Instead, as suggested in #13, the `OnChanged` callbacks should always be called when the values are changed, but that's _not_ where the config variables should be set.  Instead, they should be set only in `OnSave`

andrewgu replied:
Are you suggesting that each control have a separate OnChanged and OnSave callback?

BlueRaja replied:
Each control has its own `OnChanged` callback, and the page has the `OnSave` callback.  See #13

BlueRaja commented on ModConfigMenu/ModConfigMenu/Src/ModConfigMenuAPI/Classes/MCM_API_Setting.uc:
```diff
@@ -20,6 +18,10 @@ function name GetName();
 
+// When you mouse-over the setting.
+function SetTooltip(string Tooltip);
+function string GetTooltip();
```
:+1:

BlueRaja commented on ModConfigMenu/ModConfigMenu/Src/ModConfigMenuAPI/Classes/MCM_API_SettingsPage.uc:
```diff
@@ -6,17 +6,36 @@ function int GetPageId();
+function MCM_API_Checkbox AddCheckbox(name SettingName, string Label, string Tooltip, bool InitiallyChecked, delegate<BoolSettingHandler> Handler);
+function MCM_API_Slider AddSlider(name SettingName, string Label, string Tooltip, float SliderMin, float SliderMax, float SliderStep, float InitialValue, delegate<FloatSettingHandler> Handler);
+function MCM_API_Spinner AddSpinner(name SettingName, string Label, string Tooltip, array<string> Options, string Selection, delegate<StringSettingHandler> Handler);
+function MCM_API_Dropdown AddDropdown(name SettingName, string Label, string Tooltip, array<string> Options, string Selection, delegate<StringSettingHandler> Handler);
```
These should probably exist in a separate helper class _(something like `MCM_API_ControlFactory`)_.  Instead there should be a single `AddControl(MCM_API_Setting)` method.  This allows us (or users) to add new controls in the future, without having to change the interface for this class.

Ideally we'd just use the straight-up constructors, but that would require shipping the implementations...

andrewgu replied:
I see what you mean about having an explicit factory, but in this case the abstraction probably complicates implementation too much to be worth it. Since we have to call Spawn on the control class from inside the underlying page, we would probably have to do some sort of deferred instantiation to preserve the idea that you could create the control and then add it to any page.

Since we already have an API call for creating your own custom settings page and just hooking into the tab, I would rather save custom controls for a "version 2" extension of the API.

BlueRaja commented on ModConfigMenu/ModConfigMenu/Src/ModConfigMenuAPI/MCM_API_Includes.uci:
```diff
@@ -0,0 +1,44 @@
+`define BasicSpinnerHandler(Varname)\
+	function `GenerateHandlerName(`Varname)(MCM_API_Setting Setting, name SettingName, string SettingValue) {`Varname = SettingValue;}
+`define BasicDropdownHandler(Varname)\
+	function `GenerateHandlerName(`Varname)(MCM_API_Setting Setting, name SettingName, string SettingValue) {`Varname = SettingValue;}
```
We should preface all the macros with the same string, something like `MCM_DropdownHandler`

andrewgu replied:
Agreed.

BlueRaja commented on ModConfigMenu/ModConfigMenu/Src/ModConfigMenu/Classes/MCM_ExampleTestHarness.uc:
```diff
@@ -0,0 +1,68 @@
+	`log("MCM: Reset button clicked on page " $ string(Page.GetPageID()));
+}
+
+`MCM_API_Boilerplate(ClientModCallback);
```
Looks nice!

andrewgu commented:
@Superd22 I'm flexible, but currently there are some potentially big questions we still need to decide on, specifically how to handle save behavior. Let's revisit how we'll divvy up the work once we are at a good place with these API changes.

SaveSetting implementation: I still think it should be up to the developer how to make SaveSetting work. Since there are multiple ways to do it but most devs would probably end up calling SaveConfig on classes defined in the client mod, I think it's wiser to stay unopinionated on the matter. Best we can do would be what the Basic\* macros in this proposal do, which is to provide an easy way to write a value into a variable on change/save.

andrewgu commented:
@Superd22 Oh shit, forgot, do we still want to implement settings groups?

BlueRaja commented:
If we did, rather than baking groupings into the settings page, we could have a `GroupBox` control, which simply contains an array of other controls.  This is how every UI framework I'm aware of handles groupings/child-controls.

This was the other part of my motivation of not wanting `AddLabel`, `AddButton`, etc. in the settings page class.

andrewgu commented:
@BlueRaja Unlike every other UI framework I'm aware of, this one requires a parent control to spawn more controls instead of some process-scoped factory/manager. 

For groups: assuming we implement groupings, should they be mandatory or optional? They're quite unnecessary for small mods, but more extensive ones would obviously benefit, so I am leaning towards optional, and will propose a revision implementing optional groups unless there's a good argument for mandatory groups.

BlueRaja commented:
> this one requires a parent control to spawn more controls instead of some process-scoped factory/manager

You're conflating different issues.

What I'm suggesting would look like this:

```
local ILabel myLabel = `MCM_CreateLabel( ... );
local ICheckbox myCheckbox = `MCM_CreateCheckbox( ... );
local IGroupbox myGroupbox = `MCM_CreateGroupbox( ... );

myGroupbox.AddControl(myLabel);
myGroupbox.AddControl(myCheckbox);

MainPage.AddControl(myGroupbox);
```

Where the `MCM_CreateLabel()` etc. macros expand to something like `ControlFactory.CreateLabel()`.  If you wanted to keep those functions in `MainPage` rather than creating a new class, that's fine I guess _(though I don't think it should be the responsibility of the page to create controls)_.  But the point is that creating the control and adding it to the page should be two separate steps, so that controls can be added to other controls (the groupbox).

If you want to keep the common-case dead simple, you could have a second set of macros like

```
`define MCM_AddLabel( .. ) `{Page}.AddControl(`MCM_CreateLabel( .. ))
```

Superd22 commented:
@andrewgu i meant the `Page.ShowSettings();` implementation, not saveSettings, my bad.
To the grouping, i do lean heavily on optional, and i think those groups should be abstract to the user.

So the modder : 
1. Spawns a group like he would any other setting.
2. Simply indicate he wants a given setting to be in the group.

```
// ButtonLabel will be in GroupLabel
Page.AddGroup(GroupLabel,...)
Page.AddButton(ButtonLabel, ..., GroupLabel)
Page.AddCheckbox(CheckBoxLabel,...)
```

In this example, we can either set up the Checkbox to end-up in GroupLabel (ie once you declare a group every settings that come after are in this group by default.)
or we can set it up so CheckBoxLabel ends-up below the group in an "ungrouped" fashion.

AddGroup could take an optional ParentGroup parameter to create children.

UI wise it's really not that hard to handle in any case.

andrewgu commented:
@BlueRaja I got what you were saying. The problem is that the most granular constructor that Xcom/Unreal provides is through the Spawn function, which in this case requires a parent UIPanel instance. The problem is that a parent-less call to "MCM_CreateLabel" is potentially over-complicated because the IButton object would have to be a facade until it gets added to a Page/Group, at which point you now have a parent UIPanel instance to call Spawn on.

If either of you ( @Superd22 as well) can come up with a graceful way to work around this limitation, then I agree that it is more graceful to separate control creation from insertion. But without a workaround, it just makes implementing the actual code too complicated and not worth the tradeoff of better compliance with general UI library conventions.

BlueRaja commented:
Can't the `UIPanel` representing the control be `Spawn`ed in the `Add()` function, and stored in an instance variable?

andrewgu commented:
@BlueRaja What do you mean? As in store settings in the Create*() function and do the actual instantiation in the Add() function?

BlueRaja commented:
I mean that the `Checkbox` class does not need to extend `UIPanel`, it could just contain one.

Superd22 commented:
@BlueRaja so basically, create a dummy checkbox (not extending a panel but a dummy nonetheless).

> potentially over-complicated because the IButton object would have to be a facade until it gets added to a Page/Group, at which point you now have a parent UIPanel instance to call Spawn on.

Is it really worth-it ?

BlueRaja commented:
Is what really worth what?  Nothing is a facade of anything, there is no reason the `IButton` needs to also be a `UIPanel`.

I do believe it's worth separating the creation of the controls from adding them, at the very least to support Groupboxes.  The suggestion you made above could not support groupboxes, it could only be used for adding header text.

However, I'm not really familiar with the UI capabilities in XCOM 2.  If creating an actual groupbox-UI is not an option, and we're forced to add controls linearly one on top of the next, then header text will probably be as good as anything.

Superd22 commented:
@BlueRaja : If by GroupBoxes you only mean drawing an actual box around a few settings and performing actions on that group (toggle display, change background color, state, ect) i'm quite confident my suggestion above works like a charm. As long as you call AddGroup() before trying to populate settings in the given group. So if that's the only reason you wanna separate creation of a setting and it's insertion into the U.I you're covered.

As for the use of a "facade", you're advocating we create an object which will hold all the relevant data for a given setting, AND its UI counterpart, which will then have to be populated with that same data by that holder object. So i'm asking is it really worth going that extra step ? What will that achieve ?

**EDIT:**
In XCOM2 UI you can get a Child of a panel by its name. So this would work just fine:

```
// SettingsPage
function AddCheckBox(name label, bool value, ..., optional name group) {
local UIPanel parent; 
if(group != '') 
 parent = getChild(group);
else 
 parent = self;

Spawn(class'MCM_Checkbox', parent);
}
```

as long of course as you've already spawned a group via AddGroup()

andrewgu committed:
Added grouping mechanism.

andrewgu commented:
@Superd22 @BlueRaja 

Back from the dead (job craziness). I've modified the API to do two things, added as the latest commit to this pull request, looking for feedback:
1. Added a group mechanism to optionally create settings groups. If no group is specified, then MCM would display the setting as a "General" or "groupless" group.
2. Got rid of event mode. Settings now have both a SaveHandler and a ChangeHandler. ChangeHandler is triggered when the setting changes for any reason. SaveHandler is only triggered when attempting to save, _before_ the MCM_API_SettingsPage's Save event is triggered. Updated shorthand macros to match.

I did not prefix any of the macros yet. Good idea but it's a trivial change that doesn't require discussion, so I can always do that later.

Patrick-Seymour commented:
Hey there, andrewgu pointed me to the github for tossing in comments and ideas and such, so I thought I was throw in a few cents.

**UCI Includes**
Your proposed API changes are using a UCI file for defining preprocessor statements and such. I've actually never been able to get ModBuddy to properly compile when using UCI files, so if it is working for you, I'd love to figure out why so I can use includes in my own projects! If you're experiencing issues with UCI includes, then read on.

I think I tracked down the issue to one of the DLLs that ModBuddy uses to manage its projects. As part of the compilation process, the contents of SDK/Development/SrcOrig is copied to SDK/Development/Src, and then your project's /Src is copied, replacing anything that is modified. But that copy operation only selects files that end in the *.UC extension, so it skips the UCI files from your project's /Src directory entirely.

I've tried creating a modified DLL to fix this by having it copy UCI files as well, but the build process then complained about not being able to delete an empty directory... for some reason. If you do end up having issues with UCI includes, let me know and I'll try and help work out the issue once and for all. Of course, they may have fixed this in the month or so since I've looked at it.

**Tooltips**
So from the comment in the API_Setting.uc, it sounds like the tooltip is going to be a small popup that appears when you mouse over an option. But there is a great deal of space on the right side of the options menu (at least when you are in the MAIN main menu options).

It would be nice to be able to display detailed information if it is necessary. Some soldier skill perks or PSI abilities in-game have a little yellow question mark that you can click on to bring up a larger popup with detailed skill information. That question mark could be used in a similar manner, where the right side of the options menu displays detailed information when clicked.

Even putting one of those question marks next to the mod page label (on the left) might be nice so that mod authors can make a little About page about their mod, using the same right-side extra window.

andrewgu committed:
Partial implementation of new API. Not yet tested/verified.

andrewgu committed:
Working prototype, need to fill out the settings types and then we have 1.0.

andrewgu committed:
Finished basic 1.0 implementation, just need to test individual controls and add macros for boilerplate now.

andrewgu committed:
Fixing tab hygeine because goddamn that was ugly.

andrewgu committed:
Testing saveconfig stuff.

andrewgu committed:
Working, saving example.

andrewgu commented:
Finished mostly-done version of 1.0 API, so I'm merging.

Patrick-Seymour commented:
Groups and individual Settings are identified by names, but Pages are still identified by an ID (that can't actually be used to select the Page from the API instance object). It would be nice if Pages could be given names like Groups and Settings, and could be accessed from the API instance object. Something like:

```
ConfigAPI.NewPage('MyPage', "My Mod's Options");
...
Page.GetName(); // 'MyPage'
...
ConfigAPI.GetPage('MyPage').GetGroup('MyGroup').GetSetting('MySetting').GetValue();
```

Yeah, it means that mods would have to be careful with their page naming to prevent a collision, but it's not that hard to just recommend that mods qualify their page names, like 'MyModName_MyPage'.

Would also be interesting to see if each "level" could references its parent, so something like:

```
Setting.GetGroup().GetPage();
```

Thoughts?

andrewgu commented:
@Patrick-Seymour Good points! 

I can easily add a "find by ID" function, but will keep pages at ID's instead of names. The reason is that the list of pages is shared by all participating mods, whereas what goes into a page is owned solely by the mod itself. Naming conflicts for pages would be hard to debug because the error might only happen when a specific other mod is active. Thus I think it's a better approach to have MCM assign the ID's to make sure naming conflicts cannot happen.

But being able to query for a Page object by ID makes a lot of sense!

Separately, if you want a working example of using .uci includes with vanilla ModBuddy, the ModConfigDependencyTest project has one!

Patrick-Seymour commented:
Things are looking pretty good so far, slightly behind MASTER as of this post but it looks like nothing major changed since I downloaded it last.

https://www.dropbox.com/s/fp7xvhs97xhjqa2/mcm-implementation-20160422.png?dl=0

I've been working on this stuff for awhile, putting notes together, so some of this might not make much sense. But I got a couple issues and things to suggest:

**Issues**

If a Page has a bunch of settings, scrolling doesn't work with the mouse wheel if your mouse cursor is over a Group label. Scrolling does work if your mouse is over a Setting or the scrollbar itself. (Note that this issue only seems to present itself when the mod options are displayed as a 3D screen, scrolling seems to work fine as a 2D screen.)

The Save and Exit and Cancel buttons are selected by default when the UIScreen loads (see image). I've noticed similar issues with the Guerilla Tactics School list of Upgrade Projects: Labels appear highlighted when the menu loads, but once you mouse over-and-out of each label they appear correctly.

The Settings' callback functions only seem to be called when the page is saved, instead of when they are clicked on. The readme says "Write a handler to save the settings when they're changed", which is a bit ambiguous as to whether the callback is supposed to be called when "saved" or "changed" (clicked). But it prevents me from having dependencies in Settings, such as enabling/disabling Settings (editable) based on a checkbox. (This also results in things like dropdown selections indicating their current selection index is 0, even if changed. It's like the callbacks just arn't working.)

Tooltips don't seem to be working. I don't see tooltips on the vanilla options, even though some of them have tooltips defined using the same method you're using (BG.SetTooltipText). Maybe the vanilla tooltips are just garbage? They do work in some situations, like when you are in UIArmory_Promotion and the right-side ability summary screen shows ability tooltips when you mouse over them (defined in UIUtilities_Strategy.AddAbilitySummaryTooltip or thereabouts).

This is actually something that you might not be able to resolve without going to a 2D UIScreen, but a soldier with a big gun (say a Mag Cannon) can block-out the scrollbar or some checkboxes, making interacting with the options menu a bit hit-or-miss at times. The tactical/strategy game works fine, the main menu is just an issue because the soldier can appear "on top" of the menu data.

ESC should cancel and exit without saving changes. The vanilla options menu has a popup requesting confirmation if there are unsaved changes, but I don't think that's necessary to include if you don't.

**Suggestions**

The following could be added to MCM_API_SettingsPage, to: allow iterating of Groups without needing their names, as Groups can iterate over Settings without names so why not allow iterating of the Groups too? Would allow me to write some simple code to iterate over all Groups and Settings of a Page, and just take specific action via a switch on the Settings name only.

```
function int GetNumberOfGroups();
function MCM_API_SettingsGroup GetGroupByIndex(int i);
```

The following could be added to MCM_API_Setting and/or MCM_API_SettingsGroup, to: allow individual Settings to get their owned Group by reference, and similarly allow individual Groups to get their owned Parent by reference.

```
function MCM_API_SettingsGroup GetParentGroup();
function MCM_API_SettingsPage GetParentPage();
```

Basically simplifies stuff like this, enabling/disabling one setting dependent on the value of another setting.

```
function CheckboxCallback(MCM_API_Setting Setting) {
  local MCM_API_Checkbox CheckboxSetting;
  CheckboxSetting = MCM_API_Checkbox(Setting);
  switch (Setting.GetName()) {
    case 'EnableDisableSwitch':
      Setting.GetParentGroup().GetSettingByName('DependentSetting').SetEditable(CheckboxSetting.GetCheckboxValue());
      break;
  }
};
```

Might also be nice to have a direct Setting access on the Page itself, since Groups just add one layer of iteration necessary to find a Setting:

```
Page.GetSettingByName('SettingName');
```

Also not entirely sure why the Settings' callback handlers have SettingName as one of the fields. Isn't this just Setting.GetName()? Pretty much makes it not necessary at all to have, since you don't even have to cast to another MCM instance type to access that function. But SettingValue is nice so that I don't have to make a cast to Checkbox or Dropdown or what-have-you.

andrewgu commented:
@Patrick-Seymour 

Scrolling not working when mouse-over on a group heading: I think we'll leave that as a bug to fix in version 1.1. It's not great, but it also doesn't require API changes, and mostly works.

Save and Exit and Cancel buttons: yeah, not sure why it's happening, but again I think we can leave it as a 1.1 bug since it's small and should not require API changes, and it's not unique to this particular interface.

Settings callback functions: there are actually two that you can define. One for when it's saved, and one for when it's changed. The one you're looking for in order to manage dynamic settings is the "ChangeHandler" which is the second of the two callbacks accepted by the various "Add***" calls. For reference:

```
function MCM_API_Slider AddSlider(name SettingName, string Label, string Tooltip, float SliderMin, float SliderMax, float SliderStep, float InitialValue, 
    optional delegate<FloatSettingHandler> SaveHandler, 
    optional delegate<FloatSettingHandler> ChangeHandler);
```

`SaveHandler` is triggered only when actually "saving" but ChangeHandler triggers on every change. Let me know if that doesn't suit your needs.

Tooltips: yup, they're broken. I spent some time trying to figure out how it worked in the UIOptionsPCScreen, and I think there's some hard-coding going on in the Scaleform UI. There's a tooltip system in the UI scripts that should be usable, but it may take some time to hook that feature up. I'm inclined to leave the relevant parts in the API and make it a 1.1 feature.

Mag cannon blocking out scrollbar and checkboxes: yeah, not sure about that one. I believe the options screen has a similar problem. This might be avoidable by shifting the whole panel to the left. Again I'm inclined to make it a 1.1 thing.

In general I can be convinced otherwise about the things that I want to defer to 1.1.

ESC and popup about discarding changes: at least the popup should be there. Hooking up the ESC button should be straightforward. Will address it for 1.0.

If you wouldn't mind, can you make issues for these suggestions? They are good ones, but a pull request discussion probably isn't the best place for it.

API suggestions: they make sense. I'll change them for 1.0.

Patrick-Seymour commented:
Ahh, on-change handlers too. That explains it!

I'll make separate issues for these then, and if I come across anything else later.

andrewgu commented:
@Patrick-Seymour Just a heads up, your API suggestions have been implemented in the latest master branch. Callbacks now do not include the SettingName parameter since it's easy to get, and MCM_API_SettingsGroup and MCM_API_Setting both include a "GetParent***()" function.
